drivers/pwm/pwm_pca9685.c

```diff
@@ -0,0 +1,257 @@
+ */
+
+#define DT_DRV_COMPAT nxp_pca9685
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/pwm.h>
+#include <zephyr/drivers/pwm/pca9685.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pwm_pca9685, CONFIG_PWM_LOG_LEVEL);
+
+#define OSC_CLOCK_MAX	  50000000
+#define CHANNEL_CNT	  16
+
+#define ADDR_MODE1	  0x00
+#define SLEEP		  BIT(4)
+#define AUTO_INC	  BIT(5)
+#define RESTART		  BIT(7)
+
+#define ADDR_MODE2	  0x01
+#define OUTDRV_TOTEM_POLE BIT(2)
+#define OCH_ON_ACK	  BIT(3)
+
+#define ADDR_LED0_ON_L	  0x06
+#define ADDR_LED_ON_L(n)  (ADDR_LED0_ON_L + (4 * n))
+#define LED_FULL_ON	  BIT(4)
+#define LED_FULL_OFF	  BIT(4)
+#define BYTE_N(word, n)	  ((word >> (8 * n)) & 0xFF)
+
+#define ADDR_PRE_SCALE	  0xFE
+
+/*
+ * Time allowed for the oscillator to stabilize after the PCA9685's
+ * RESTART mode is used to wake the chip from SLEEP.
+ * See PCA9685 datasheet section 7.3.1.1
+ */
+#define OSCILLATOR_STABILIZE K_USEC(500)
+
+struct pca9685_config {
+	struct i2c_dt_spec i2c;
+	bool outdrv_open_drain;
+	bool och_on_ack;
+	bool invrt;
+};
+
+struct pca9685_data {
+	struct k_mutex mutex;
+	uint8_t pre_scale;
+};
+
+static int set_reg(const struct device *dev, uint8_t addr, uint8_t value)
+{
+	const struct pca9685_config *config = dev->config;
+	uint8_t buf[] = {addr, value};
+	int ret;
+
+	ret = i2c_write_dt(&config->i2c, buf, sizeof(buf));
+	if (ret != 0) {
+		LOG_ERR("I2C write [0x%02X]=0x%02X: %d", addr, value, ret);
+	} else {
+		LOG_DBG("[0x%02X]=0x%02X", addr, value);
+	}
+	return ret;
+}
+
+static int get_reg(const struct device *dev, uint8_t addr, uint8_t *value)
+{
+	const struct pca9685_config *config = dev->config;
+	int ret;
+
+	ret = i2c_write_read_dt(&config->i2c, &addr, sizeof(addr), value, sizeof(*value));
+	if (ret != 0) {
+		LOG_ERR("I2C write [0x%02X]=0x%02X: %d", addr, *value, ret);
+	}
+	return ret;
+}
+
+static int set_pre_scale(const struct device *dev, uint8_t value)
+{
+	struct pca9685_data *data = dev->data;
+	uint8_t mode1;
+	int ret;
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	/* Unblock write to PRE_SCALE by setting SLEEP bit to logic 1 */
+	ret = set_reg(dev, ADDR_MODE1, AUTO_INC | SLEEP);
+	if (ret != 0) {
+		goto out;
+	}
+
+	ret = get_reg(dev, ADDR_MODE1, &mode1);
+	if (ret != 0) {
+		goto out;
+	}
+
+	if ((mode1 & RESTART) == 0x00) {
+		LOG_ERR("RESTART bit should be set");
+		return -EIO;
+	}
+
+	ret = set_reg(dev, ADDR_PRE_SCALE, value);
+	if (ret != 0) {
+		goto out;
+	}
+
+	/* Clear SLEEP bit - See datasheet section 7.3.1.1 */
+	ret = set_reg(dev, ADDR_MODE1, AUTO_INC);
+	if (ret != 0) {
+		goto out;
+	}
+
+	k_sleep(OSCILLATOR_STABILIZE);
+
+	ret = set_reg(dev, ADDR_MODE1, AUTO_INC | RESTART);
+	if (ret != 0) {
+		goto out;
+	}
+
+out:
+	k_mutex_unlock(&data->mutex);
+
+	return ret;
+}
+
+static int pca9685_set_cycles(const struct device *dev, uint32_t channel, uint32_t period_count,
+			      uint32_t pulse_count, pwm_flags_t flags)
```

Comment:
Can you please follow the "preferred limit on the length of a single line" coding style everywhere, and not just blindly apply clang-format?

```c
static int pca9685_set_cycles(const struct device *dev,
			      uint32_t channel, uint32_t period_count,
			      uint32_t pulse_count, pwm_flags_t flags)
```

Reply:
Done

Reply:
Actually not, please apply "preferred limit on the length of a single line" coding style everywhere it makes sense.

Reply:
Apologies I missed that, can you please check again @jfischer-no